### PR TITLE
xFormer updates to fMHA FW

### PR DESCRIPTION
**Pull request by danthe3rd**
Note: this changes the input/output formats to BMHK from BHMK (see https://github.com/NVIDIA/cutlass/issues/771). The output will be contiguous (no longer possible to specify custom strides for the outputs)

I plan to open another PR after this one to bring the changes from https://github.com/facebookresearch/xformers/pull/587 once it's merged

(A lot of the changes touch files also used for the BW pass, which will be upstreamed later, but I prefer to keep everything in sync as it's easier)

### CODE REVIEW

**Comment by hwu36** [comment]
@terrychenism, i assume the change of `input/output formats to BMHK from BHMK` is okay to you since AIT is already doing it.

@mnicely , who else in addtion to ait uses xformers mha through cutlass?  we need to notice them, too.

This api change is needed for the issue mentioned in #771

**Comment by terrychenism** [comment]
Yeah, thanks for the updates! It's aligned with AIT.

**Comment by danthe3rd** [comment]
@terrychenism as it is right now (as in xFormers), the output will always be contiguous with shape BMHK. Is there a use-case where you need to have custom strides for the output as well?

**Review by terrychenism — APPROVED** [review]

File: examples/41_fused_multi_head_attention/mma_from_smem.h [hunk 1]
```diff
@@ -57,8 +57,8 @@
 #include "epilogue_thread_apply_logsumexp.h"
 #include "gemm_kernel_utils.h"
 #include "iterators/make_residual_last.h"
-
-/////////////////////////////////////////////////////////////////////////////////////////////////
+#include "iterators/transpose_warp_iterator.h"
+#include "iterators/warp_iterator_from_smem.h"
```
Comment (tianleiwu) [inline examples/41_fused_multi_head_attention/mma_from_smem.h:61]: Need add these files to avoid build error.

Comment (danthe3rd) [reply examples/41_fused_multi_head_attention/mma_from_smem.h:61]: Oh my bad! Adding them right now

**Comment by tianleiwu** [comment]
@danthe3rd, there is a line in kernel_forward.h is nice to change.

When an application is compiled with sm from 37 to 80 ( to support as many device as possible), this line will cause build error:
https://github.com/NVIDIA/cutlass/blob/0cc33af50c9667c57d842f92512f4cfc3dcfd6d4/examples/41_fused_multi_head_attention/kernel_forward.h#L924
It prevents us to directly use kernel_forward.h. Users like us have to patch the line.

Proposed change is to comment the line like in pytorch:
https://github.com/pytorch/pytorch/blob/95e3e339a8f896799872fad910fc60ea13d96ead/aten/src/ATen/native/transformers/cuda/mem_eff_attention/kernel_forward.h#L878

**Comment by danthe3rd** [comment]
> @danthe3rd, there is a line in kernel_forward.h is nice to change.

Good catch! I removed that

File: examples/41_fused_multi_head_attention/fused_multihead_attention_fixed_seqlen.cu [hunk 2]
```diff
@@ -839,18 +857,17 @@ public:
...
-      p.v_strideH = p.v_strideM * options.seq_length_kv;
-      p.o_strideH = options.head_size_v * options.seq_length;
-      p.q_strideB = p.q_strideH * options.head_number;
-      p.k_strideB = p.k_strideH * options.head_number;
-      p.v_strideB = p.v_strideH * options.head_number;
-      p.o_strideB = options.head_size_v * options.seq_length * options.head_number;
+      p.q_strideB = p.q_strideM * options.seq_length;
+      p.k_strideB = p.k_strideM * options.seq_length_kv;
+      p.v_strideB = p.v_strideM * options.seq_length_kv;
+      std::cout << "stridesQ: " << p.q_strideB << ", " << p.q_strideM << ", " << p.q_strideH << std::endl;
```
Comment (MARD1NO) [inline examples/41_fused_multi_head_attention/fused_multihead_attention_fixed_seqlen.cu:870]: Maybe you should delete this debug code?

**Comment by MARD1NO** [comment]
Hey guys, Now we change the format from BHMK to BMHK，and in `check_supported`, we still check `stride_M % kAlignment`，shall we change to check stride_H？

File: examples/41_fused_multi_head_attention/fused_multihead_attention_fixed_seqlen.cu (hunk 2 above)
Comment (danthe3rd) [reply examples/41_fused_multi_head_attention/fused_multihead_attention_fixed_seqlen.cu]: Good point - removed thanks :)

**Comment by danthe3rd** [comment]
> Hey guys, Now we change the format from BHMK to BMHK，and in `check_supported`, we still check `stride_M % kAlignment`，shall we change to check stride_H？ @danthe3rd

We check both `strideM` and `strideH` which I believe is correct

**Comment by MARD1NO** [comment]
> > Hey guys, Now we change the format from BHMK to BMHK，and in `check_supported`, we still check `stride_M % kAlignment`，shall we change to check stride_H？ @danthe3rd
> 
> We check both `strideM` and `strideH` which I believe is correct

sry, I missed these code... :)

**Review by hwu36 — APPROVED** [review]

**Pull request merged**